src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs

```diff
@@ -422,5 +423,59 @@ private UnboundLambda BindAnonymousFunction(AnonymousFunctionExpressionSyntax sy
 
             return lambda;
         }
+
+        // Please don't use thread local storage widely. This should be one of only a few uses.
+        [ThreadStatic] private static PooledDictionary<SyntaxNode, int>? s_lambdaBindings;
+
+        internal TResult BindWithLambdaBindingCountDiagnostics<TSyntax, TArg, TResult>(
+            TSyntax syntax,
+            TArg arg,
+            BindingDiagnosticBag diagnostics,
+            Func<Binder, TSyntax, TArg, BindingDiagnosticBag, TResult> bind)
+            where TSyntax : SyntaxNode
+            where TResult : BoundNode
+        {
+            Debug.Assert(s_lambdaBindings is null);
+            s_lambdaBindings = PooledDictionary<SyntaxNode, int>.GetInstance();
+
+            try
+            {
+                TResult result = bind(this, syntax, arg, diagnostics);
+
+                foreach (var pair in s_lambdaBindings)
```

Comment:
>s_lambdaBindings [](http://example.com/codeflow?start=37&length=16)

Not sure how slow is the read of TLS. Since we pre-allocate the dictionary now, consider storing its reference in a local and using the local throughout of the method. Limiting assess to ```s_lambdaBindings``` to changing its value and communiacting across method boundaries.